### DISCUSSION TITLE
Set-TSSSecret -Folder Option Not Working?

### DISCUSSION THREAD

**Discussion by Thranx** [thycotic-ps/thycotic.secretserver#104] [Q&A : Module] [answered]
Edit:  End result, Set-TSSSecret -Folder option only works in Secret Server 10.9 and newer. 

-----------

Maybe I'm using this wrong, or maybe my expectations for this switch are inaccurate, but passing the folder ID to the -Folder switch in Set-TSSSecret.  Example: 

`Set-TssSecret -TssSession $session -SecretId 473 -Folder 37`

The error returned is a 404.

```
Set-TssSecret : 
404 - File or directory not found.
Server Error

  404 - File or directory not found.
  The resource you are looking for might have been removed, had its name changed, or is temporarily unavailable.
```

Am I using this wrong?  What's the proper way to move a Secret/change the folder value?  The documentation for -Folder doesn't actually appear to be for -Folder.  (https://thycotic-ps.github.io/thycotic.secretserver/commands/Set-TssSecret#folder)

**Comment by wsmelton** [comment, marked as answer]
If you run the command with `-Verbose` you can verify the output and the path being used along with the payload provided. Using `-Folder` on the command will move the provided secret ID(s) to that folder:

![image](https://user-images.githubusercontent.com/11204251/109577165-c0acfc00-7aba-11eb-919d-808a94f6f2db.png)

This is run against a 10.9.37 build of Secret Server.

**Reply by Thranx** [reply]
![image](https://user-images.githubusercontent.com/19175599/109691136-f9d98080-7b3b-11eb-8547-418334c623e1.png)


Nothing stands out.  I if do the same operation on a field like notes for that same secret, it changes the value fine.

I'm on Secret Server 10.7, and I tried with module versions 0,27, 0.32 and 0.34.

**Reply by wsmelton** [reply]
After the failure, if you can look at `$error[0] | Select *` and see if that gives any additional details in the output.

**Reply by Thranx** [reply]
This is error[0]

```
PSMessageDetails      :
Exception             : System.ArgumentException: Invalid JSON primitive:  File or directory not found.





                        Server Error


                          404 - File or directory not found.
                          The resource you are looking for might have been removed, had its name changed, or is temporarily unavailable.




                        .
                           at System.Web.Script.Serialization.JavaScriptObjectDeserializer.BasicDeserialize(String input, Int32 depthLimit, JavaScriptSerializer
                        serializer)
                           at System.Web.Script.Serialization.JavaScriptSerializer.Deserialize(JavaScriptSerializer serializer, String input, Type type, Int32
                        depthLimit)
                           at Microsoft.PowerShell.Commands.JsonObject.ConvertFromJson(String input, ErrorRecord& error)
                           at Microsoft.PowerShell.Commands.ConvertFromJsonCommand.ConvertFromJsonHelper(String input)
                           at System.Management.Automation.CommandProcessorBase.Complete()
TargetObject          :
CategoryInfo          : NotSpecified: (:) [ConvertFrom-Json], ArgumentException
FullyQualifiedErrorId : System.ArgumentException,Microsoft.PowerShell.Commands.ConvertFromJsonCommand
ErrorDetails          :
InvocationInfo        : System.Management.Automation.InvocationInfo
ScriptStackTrace      : at <ScriptBlock><Process>, C:\Users\--username--\Documents\WindowsPowerShell\Modules\Thycotic.SecretServer\parts\ErrorHandling.ps1: line 9
                        at Set-Secret<Process>, C:\Users\--username--\Documents\WindowsPowerShell\Modules\Thycotic.SecretServer\functions\secrets\Set-Secret.ps1: line 561
                        at <ScriptBlock>, <No file>: line 1
PipelineIterationInfo : {}
```

**Reply by wsmelton** [reply]
If you can run the API call manually using something like below (based on your screenshot)

```powershell
$object = @{
	"data" = @{
		"folder" = @{
			"dirty" = $true,
			"value" = 37
		}
	}
} | ConvertTo-Json

$uri = 'https://.../SecretServer/api/v1/secrets/473/general'

Invoke-TssRestApi -Uri $uri -PersonalAccessToken $session.AccessToken -Method PATCH -Body $object
```

**Reply by Thranx** [reply]
Heh, I was just trying to figure out how to use the invoke-tssrestapi and was hung up on the personal access token switch.

Tried it.  Same error.  Note, the comma after $true had to be switched to a semi-colon.

Thanks for helping on this buy the way.  I have a sandbox I can try and upgrade to 10.9 in a bit here if you think that could be the issue.

```
Invoke-RestMethod : 
404 - File or directory not found.
Server Error
 
  404 - File or directory not found.
  The resource you are looking for might have been removed, had its name changed, or is temporarily unavailable.
 
At C:\Users\--username--\Documents\WindowsPowerShell\Modules\Thycotic.SecretServer\functions\Invoke-RestApi.ps1:135 char:9
+         Invoke-RestMethod @irmSplat |
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-RestMethod], WebException
    + FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeRestMethodCommand
```

```
PSMessageDetails      :
Exception             : System.Net.WebException: The remote server returned an error: (404) Not Found.
                           at Microsoft.PowerShell.Commands.WebRequestPSCmdlet.GetResponse(WebRequest request)
                           at Microsoft.PowerShell.Commands.WebRequestPSCmdlet.ProcessRecord()
TargetObject          : System.Net.HttpWebRequest
CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-RestMethod], WebException
FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeRestMethodCommand
ErrorDetails          :



                        404 - File or directory not found.





                        Server Error


                          404 - File or directory not found.
                          The resource you are looking for might have been removed, had its name changed, or is temporarily unavailable.





InvocationInfo        : System.Management.Automation.InvocationInfo
ScriptStackTrace      : at Invoke-RestApi<Process>, C:\Users\--username--\Documents\WindowsPowerShell\Modules\Thycotic.SecretServer\functions\Invoke-RestApi.ps1: line 135
                        at <ScriptBlock>, <No file>: line 1
PipelineIterationInfo : {}
```

**Reply by wsmelton** [reply]
Can you actually move a secret via the UI?

**Reply by Thranx** [reply]
Ha, great test.  Yea, it moves fine via the UI.  (edit: using the same account/credentials... so not a permissions issue)

**Reply by wsmelton** [reply]
http://gph.is/1T9w5OJ

Yep, forgot that the `PATCH` method was not added until the 10.9.0 release.

**Reply by Thranx** [reply]
Ah crud.  Any way to do it via invoke-tssrestapi and a post?  If not, I'll start working on the upgrade.  Needs to happen eventually anyway.

**Reply by wsmelton** [reply]
Yeah, `Invoke-TssRestApi` can be used I'm using PATCH endpoints as they are simpler to build the payload and call itself. You can still use the `PUT /secrets/{id}` for moving a folder, that was the method to use prior to the PATCH endpoints being added.

**Reply by Thranx** [reply]
Okay, got my sandbox and my prod updated to 10.9.  Works as expected now.

Thanks again for your help on this.
